scripts/ZoneMinder/lib/ZoneMinder/Control/Dahua.pm

```diff
@@ -0,0 +1,366 @@
+package ZoneMinder::Control::Dahua;
+
+use 5.8.0;
+use strict;
+use warnings;
+
+require ZoneMinder::Base;
+require ZoneMinder::Control;
+
+our @ISA = qw(ZoneMinder::Control);
+
+our $REALM = '';
+our $USERNAME = '';
+our $PASSWORD = '';
+our $ADDRESS = '';
+our $PROTOCOL = 'http://';
+
+use Time::HiRes qw(usleep);
+
+use ZoneMinder::Logger qw(:all);
+use ZoneMinder::Config qw(:all);
+use ZoneMinder::Database qw(zmDbConnect);
+
+sub new
+{
+    my $class = shift;
+    my $id = shift;
+    my $self = ZoneMinder::Control->new( $id );
+    bless( $self, $class );
+    srand( time() );
+    return $self;
+}
+
+our $AUTOLOAD;
+
+sub AUTOLOAD
+{
+    my $self = shift;
+    my $class = ref($self) || croak( "$self not object" );
+    my $name = $AUTOLOAD;
+    $name =~ s/.*://;
+    if ( exists($self->{$name}) )
+    {
+        return( $self->{$name} );
+    }
+    Fatal( "Can't access $name member of object of class $class" );
+}
+
+sub open
+{
+    my $self = shift;
+    $self->loadMonitor();
+
+    # The Dahua camera firmware API supports the concept of having multiple
+    # channels on a single IP controller.
+    # As most cameras only have a single channel, and there is no similar
+    # information model in Zoneminder, I'm hardcoding the first and default
+    # channel "0", here.
+    $self->{dahua_channel_number} = "0";
+
+    my ($protocol, $username, $password, $address)
+       = $self->{Monitor}->{ControlAddress} =~ /^(https?:\/\/)?([^:]+):([^\/@]+)@(.*)$/;
```

Comment:
That's really good feedback. I guess there is no validation on the stored ControlAddress (since there can be so many applicable protocol schemes).

I used your scheme from Trendnet.pm in Dahua.pm. I'll also do the same change for #2191

Also, I removed some trailing whitespace from Trendnet.pm